01-resources-v2/proposal.md

```diff
@@ -51,11 +51,9 @@ type alias Bits = ()
 ## Versioned Artifacts interface
 
 ```elm
-spaces  : Config -> Set Space
-check   : Config -> Space -> Maybe Version -> List (Version, Metadata)
+check   : Config -> Dict Space Version -> Dict Space (List (Version, Metadata))
 get     : Config -> Space -> Version -> Bits
-put     : Config -> Bits -> (Space, Set Version)
-destroy : Config -> Bits -> Dict Space (Set Version)
+put     : Config -> Bits -> Dict Space { created : Set Version, deleted : Set Version }
```

Comment:
A thought that occurred to me here, is that in the context of a `git` resource I think it is more likely that a `get` (with `version: every`) will discover that a version is missing after force push, than `check` or `put` (which will not support deleting refs anyway)?

Might it be an idea to allow get to signal that the version being fetched no longer exists, rather than just exit with an error (if someone has force pushed pushed after the build was started)?

Reply:
If it makes sense to allow `get` to signal if a version is missing, I'm wondering if it also would make sense (since `get` would not have deleted anything itself) to just rename the keys for the returned objects to avoid mistakes on the part of the resource implementors:

- `versions` and `created` renamed to `added`.
- `deleted` renamed to `removed`

So `check` would return: 
```json
{
  "space": "master", 
  "added": [{ 
    "ref": "sha", 
    "metadata": <map>
  }]
}
```
And `get` _could_ return:
```json
{
  "space": "master", 
  "removed": [{ 
    "ref": "sha"
  }]
}
```
While `put` would return (removed optional):
```json
{
  "space": "master", 
  "added": [{ 
    "ref": "sha"
  }],
  "removed": [{ 
    "ref": "sha"
  }]
}
```

Might need better naming, but I think it might be the case that `get` will be better at discovering missing versions in some cases. E.g., with the git resource on `version: every` there might already be a later version than the one currently being passed to `get`, in which case I'd like something to let `check` know that we need to start fresh?

Reply:
Hmm, I like the symmetry here but I could see a few problems with it.

For Git repos at least, the `check` should notice a force-push pretty quickly because the latest version's `ref` should no longer be valid.

There could also be a tricky scenario where an older version of a resource that the build is trying to fetch is temporarily unavailable. Marking it 'deleted' would prevent another build from attempting to fetch it. I suppose `check` could re-discover it, but only if it's the latest version.